From ac404e78eb2c1ab030b8e0639bedb7ad979925a9 Mon Sep 17 00:00:00 2001
From: Saurabh Sengar <saurabh.singh@xilinx.com>
Date: Tue, 8 May 2018 16:59:09 +0530
Subject: [PATCH 0353/1852] drm: xlnx: mixer: Added the atomic check for
 invalid coordinates

commit dc8ee1701dc26e15a2f6b2ac6f532736d6f7d91c from
https://github.com/Xilinx/linux-xlnx.git

If application tries to set coordinates for overlay planes which are
not valid ie this overlay plane is violating the primary layer
boundaries, then this fix will return -EINVAL to userspace application
instead of failing silently.

Signed-off-by: Saurabh Sengar <saurabh.singh@xilinx.com>
Reviewed-by: Vishal Sagar <vishal.sagar@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_mixer.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_mixer.c b/drivers/gpu/drm/xlnx/xlnx_mixer.c
index 49c896c5ed75..e14f29c3c90f 100644
--- a/drivers/gpu/drm/xlnx/xlnx_mixer.c
+++ b/drivers/gpu/drm/xlnx/xlnx_mixer.c
@@ -1739,7 +1739,17 @@ static void xlnx_mix_plane_cleanup_fb(struct drm_plane *plane,
 static int xlnx_mix_plane_atomic_check(struct drm_plane *plane,
 				       struct drm_plane_state *state)
 {
-	return 0;
+	int scale;
+	struct xlnx_mix_plane *mix_plane = to_xlnx_plane(plane);
+	struct xlnx_mix_hw *mixer_hw = to_mixer_hw(mix_plane);
+
+	scale = xlnx_mix_get_layer_scaling(mixer_hw,
+					   mix_plane->mixer_layer->id);
+	if (is_window_valid(mixer_hw, state->crtc_x, state->crtc_y,
+			    state->src_w >> 16, state->src_h >> 16, scale))
+		return 0;
+
+	return -EINVAL;
 }
 
 static void xlnx_mix_plane_atomic_update(struct drm_plane *plane,
-- 
2.31.1

